### PR TITLE
Dev #18040: Add buttons submit the whole labelset instead of only add…

### PR DESCRIPTION
**Pull request by gabrieljenik**

### CODE REVIEW

**Comment by gabrieljenik** [comment]
This is ready for reviewal.

File: application/controllers/admin/Labels.php [hunk 1]
```diff
@@ -740,4 +741,64 @@ protected function renderWrappedTemplate($sAction = 'labels', $aViewUrls = array
...
+     * Called by ajax from the label set editor.
+     * @param int $lid
+     * @param string $newId
+     * @param string $code
+     * @param int $assessmentValue
+     * @param string $title
+     */
+    public function getLabelRowForAllLanguages($lid, $newId, $code, $assessmentValue = 0, $title = '')
+    {
+        if (!Permission::model()->hasGlobalPermission('labelsets', 'update')) {
```
Comment (olleharstedt) [inline application/controllers/admin/Labels.php:757]: Read permission should be enough, if there's no write in the method?

File: assets/scripts/admin/labels.js [hunk 2]
```diff
@@ -227,60 +244,75 @@ function get_next_code(event) {
...
+            // arrayofhtml is a json object containing the different HTML row by language
+            // eg: {"en":"{the html of the en row}", "fr":{the html of the fr row}}
+            $defer.resolve({ arrayofhtml: arrayofhtml });
+        },
+        error(html, status) {
+            alert('Internal error retrieving row html: ' + errormessage);
+            $defer.reject([html, status, errormessage]);
+        },
+    });
+    return $defer.promise();
```
Comment (olleharstedt) [inline assets/scripts/admin/labels.js:273]: I think you can just `return $ajax({ ...` too

File: application/controllers/admin/Labels.php (hunk 1 above)
Comment (gabrieljenik) [reply application/controllers/admin/Labels.php:757]: This is because this function is called in the process of editing labels.

Comment (olleharstedt) [reply application/controllers/admin/Labels.php:757]: You shouldn't lock a function to its context, tho, you might prevent reuse that way.

Comment (gabrieljenik) [reply application/controllers/admin/Labels.php:757]: But this is an action in the controller. Not expecting to reuse this.

Comment (Shnoulle) [reply application/controllers/admin/Labels.php:757]: > But this is an action in the controller. Not expecting to reuse this.


You can get the json just to show it to user . It's a public function. 
Why disallow show lable if user have read rights ?

This public function didn't update label : then need only read.

Comment (olleharstedt) [reply application/controllers/admin/Labels.php]: You're not expecting to reuse it _now_, but who knows what will happen 2-5 years from now.

Comment (gabrieljenik) [reply application/controllers/admin/Labels.php]: Do you expect to reuse controller actions?
Anyway, already changed the permission to read :)

Comment (olleharstedt) [reply application/controllers/admin/Labels.php]: "getLabelRowForAllLanguages" sounds like something that could be used in, say, question editor. Na ja. Thanks!

File: application/controllers/admin/Labels.php [hunk 3]
```diff
@@ -740,4 +741,64 @@ protected function renderWrappedTemplate($sAction = 'labels', $aViewUrls = array
...
+     * @param string $title
+     * @return string   The html of the row
+     */
+    private function getLabelRow($language, $first, $newId, $code, $assessmentValue, $title)
+    {
+        $view = 'labelRow.twig';
+        $aData = array(
+            'language' => $language,
+            'first' => $first,
+            'sortOrder' => $newId,
```
Comment (olleharstedt) [inline application/controllers/admin/Labels.php:792]: sort order = new id? Variable naming mismatch?

**Comment by olleharstedt** [comment]
Meh. The broken CI comes from dev branch, right?

**Comment by gabrieljenik** [comment]
> 25 changed files?

Ooops. Files from T206 got into here. Will clean up

**Comment by gabrieljenik** [comment]
Conflicts solved. Ready for re-re-review.

**Comment by Shnoulle** [comment]
re-re-re-re-re-re-re-re-review ;)

**Comment by olleharstedt** [comment]
Dev branch should be green now, can you merge upstream?

**Comment by gabrieljenik** [comment]
> can you merge upstream

What do you mean? Sorry, didn't get it. From where to where?

**Comment by olleharstedt** [comment]
Merge develop branch into this again, to get a green CI.

**Comment by gabrieljenik** [comment]
> Merge develop branch into this again, to get a green CI.

What about now?

File: application/controllers/admin/Labels.php [hunk 4]
```diff
@@ -737,4 +738,64 @@ protected function renderWrappedTemplate($sAction = 'labels', $aViewUrls = array
...
+     * @param string $title
+     * @return string   The html of the row
+     */
+    private function getLabelRow($language, $first, $newId, $code, $assessmentValue, $title)
+    {
+        $view = 'labelRow.twig';
+        $aData = array(
+            'language' => $language,
+            'first' => $first,
+            'rowId' => $newId,
```
Comment (olleharstedt) [inline application/controllers/admin/Labels.php:789]: rowId or newId?

File: assets/scripts/admin/labels.js [hunk 5]
```diff
@@ -110,6 +111,9 @@ function quickaddfunction() {
     }
 
     console.log('lsrows', lsrows);
+
+    // parsedRows is an array of objects with the parsed row data (ie: [{code:"code", titles:{"en":"Title"}}])
+    let parsedRows = [];
```
Comment (olleharstedt) [inline assets/scripts/admin/labels.js:116]: Nice with example data in comment. Maybe we can even create types here one day (TypeScript or Flow, for example).

File: application/controllers/admin/Labels.php (hunk 4 above)
Comment (gabrieljenik) [reply application/controllers/admin/Labels.php:789]: $newId is the name of the argument. Should we change it you think?

**Comment by gabrieljenik** [comment]
The "need update by author" is because of this?
https://github.com/LimeSurvey/LimeSurvey/pull/2379#discussion_r1093453153

Not sure what am I suppose to focus on :)

**Pull request merged**